sdk/src/test/java/com/google/cloud/dataflow/sdk/util/common/CounterSetTest.java

```diff
@@ -222,4 +226,110 @@ public void testAddCounterMutatorAddMultipleCounters() {
     assertThat(set, containsInAnyOrder(c1, c2));
   }
 
+  @Test
+  public void testHandlesDuplicateEntries() {
+    final String counterName = "counter-name";
+    CounterSet counters = new CounterSet();
+
+    // getting a nonexistent counter returns null
+    assertThat(counters.getExistingCounter(counterName), is(nullValue()));
+
+    // adding an already existing counter using addNewCounter throws an exception
+    Counter<Integer> counterAddCheck = Counter.ints(counterName, SUM);
+    boolean exceptionThrown = false;
+    try {
+      counters.addNewCounter(counterAddCheck);
+    } catch (IllegalArgumentException e) {
+      exceptionThrown = true;
+    }
+    assertThat(exceptionThrown, is(false));
+
+    try {
+      counters.addNewCounter(counterAddCheck);
+    } catch (IllegalArgumentException e) {
+      exceptionThrown = true;
+    }
+    assertThat(exceptionThrown, is(true));
+    assertThat(counters.getExistingCounter(counterName), is(notNullValue()));
+  }
+
+  @Test
+  public void testMutatorSanity() {
+    final String counterName = "counter-name";
+    NameContext context = new NameContext("original_name", "optimized_name");
+    Counter.Name structuredName = new Counter.StructuredNameOriginal(counterName, context);
+
+    CounterSet counters = new CounterSet();
+
+    // Create an unstructured counter with a value
+    Counter<Integer> counterUnstructured = Counter.ints(counterName, SUM);
+    counters.addNewCounter(counterUnstructured);
+    counterUnstructured.addValue(10);
+
+    // create a counter with the same name, but a different context and value
+    CounterSet.AddCounterMutator adder = counters.getAddCounterMutator(context);
+    @SuppressWarnings("unchecked")
+    Counter<Integer> counterStructured =
+        adder.addCounter(Counter.ints(counterName, SUM), Counter.NameType.ORIGINAL);
+    counterStructured.addValue(-10);
+
+    // now read back both counters and check values
+    @SuppressWarnings("unchecked")
+    Counter<Integer> counterUnstructuredToo =
+        (Counter<Integer>) counters.getExistingCounter(counterName);
+
+    @SuppressWarnings("unchecked")
+    Counter<Integer> counterStructuredToo =
+        (Counter<Integer>) counters.getExistingCounter(structuredName);
+    assertThat(counterUnstructuredToo.getAggregate(), is(10));
+    assertThat(counterStructuredToo.getAggregate(), is(-10));
+
+    // grab a regular mutator, first check that adding an existing counter simply reuses
+    adder = counters.getAddCounterMutator();
+    int size = counters.size();
+    counterUnstructuredToo = adder.addCounter(counterUnstructuredToo);
+    assertThat(counters.size(), is(size));
+    assertThat(counterUnstructuredToo.getAggregate(), is(10));
+
+    // This counter should still be unstructured, ie the two mutators should not interfere with one
+    // another
+    assertThat(counterUnstructuredToo.getNameType(), is(Counter.NameType.UNSTRUCTURED));
+  }
+
+  @Test
+  public void testEquivalentAdd() {
```

Comment:
Update name to reflect what this is actually testing (and/or split this into multiple cases)